docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,128 @@
+
+MeshService object contains VIPs in status.
+```yaml
+status: # managed by CP. Not shared cross zone, but synced to global
+  vips:
+  - ip: <kube_cluster_ip> # or kuma VIP
+    type: Kubernetes # | Kuma
+```
+
+If MeshService was created based on Kubernetes Service, we can just take ClusterIP, but otherwise we need to assign each MeshService a Kuma VIP.
+Kuma VIP is a non-routable IP used only by Envoy to indicate destination (like a host header, but on L4). We currently do this with 240.0.0.0/4 CIDR.
+
+Kuma VIPs are assigned asynchronously with MeshService creation. Assigning Kuma VIPs as sync operation on MeshService creation would create a race condition on VIP allocation.
+
+VIP assigning process (single-threaded):
+* List MeshServices
+* If a MeshService has no VIP, assign a Kuma VIP. Kubernetes Services would have a VIP already (ClusterIP), so we don’t need to assign an extra VIP.
+* To assign a VIP, we build an IPAM state reserving every address we already assigned and allocate the next address.
+  To solve the problem of “assigning immediately” a VIP of an old service to a new Service we can also store in memory recently removed VIPs for X seconds and set TTL of DNS responses for X seconds.
+  The current value of 30s is too long, especially that the DNS server is run locally in kuma-dp. Leader election should take more than X seconds, so if we were to lose this list, it’s not a problem.
+
+Following this process, we don’t have to store VIPs in any central place like we do currently (VIP ConfigMap for each Mesh).
+We also assign VIPs for ExternalServices, but this can work as long as we consider both ExternalServices and MeshServices for this process.
+
+To avoid collision with an existing VIPs mechanism, we can take a completely new CIDR (241.0.0.0/4).
+
+It's not expected for service to have multiple VIPs, but we want expose this possibility for the future (therefore `vips` field is a list).
+
+### Hostname Generator
+
+MeshService has to be addressable with a hostname.
+In case of services in a Kubernetes cluster, we don't have to do anything because we can just rely on Kube DNS that resolves to a Kube VIP we placed for MeshService.
+
+In case of services in a Universal cluster, currently we provide two ways:
+* `<kuma.io/service>.mesh` address that resolves to Kuma VIP
+* VirtualOutbound resource to provide custom addresses
+
+Hostnames should be managed only by MeshOperator. Otherwise, we get into security issues such as one service owner can override a hostname of other service.
+Hostnames are created by convention in the majority of cases - Kubernetes hostnames are also such example.
+It is still possible for service owner to create unique hostname for their app by asking reaching out to mesh operator.
+
+To manage hostname, we introduce new object called `HostnameGenerator`
+
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
```

Comment:
There can be a mesh here?

Reply:
I'm not sure. We don't have a mechanism yet to have a resource that can optionally belong to a mesh. For example, in case of Secrets we created Secret (mesh-bound) and GlobalSecret (not mesh-bound).

I think we need to decide on one way or another. I think I'd go with "global" and then extend this to be optionally "mesh-bound" if needed.

WDYT?

Reply:
Sounds good but then let's name it `GlobalHostnameGenerator` should like k8s convention of Cluster scoped resources should be named `Cluster...`

Reply:
I don't know but It's nice in k8s to have a systematic way to understand if something is cluster scoped or not.

In anycase I'd stay away from GlobalHostnameGenerator for the multizone service as it's confusing with GlobalSecret